zkstd/src/circuit/gadget/curve.rs

```diff
@@ -2,51 +2,74 @@ use super::binary::BinaryAssignment;
 use super::field::FieldAssignment;
 
 use crate::circuit::CircuitDriver;
-use crate::common::{BNProjective, CurveGroup, Group, IntGroup, Ring};
+use crate::common::{BNAffine, BNProjective, PrimeField};
 use crate::r1cs::R1cs;
 
 #[derive(Clone)]
-pub struct PointAssignment<C: CircuitDriver> {
-    x: FieldAssignment<C>,
-    y: FieldAssignment<C>,
-    z: FieldAssignment<C>,
+pub struct PointAssignment<F: PrimeField> {
+    x: FieldAssignment<F>,
+    y: FieldAssignment<F>,
+    z: FieldAssignment<F>,
 }
 
-impl<C: CircuitDriver> PointAssignment<C> {
-    pub fn instance(cs: &mut R1cs<C>, x: C::Scalar, y: C::Scalar, is_infinity: bool) -> Self {
-        let x = FieldAssignment::instance(cs, x);
-        let y = FieldAssignment::instance(cs, y);
+impl<F: PrimeField> PointAssignment<F> {
+    pub fn instance<C: CircuitDriver<Scalar = F>>(
+        cs: &mut R1cs<C>,
+        point: impl BNAffine<Base = F>,
+    ) -> Self {
+        let x = FieldAssignment::instance(cs, point.get_x());
+        let y = FieldAssignment::instance(cs, point.get_y());
         let z = FieldAssignment::instance(
             cs,
-            if is_infinity {
-                C::Scalar::zero()
+            if point.is_identity() {
+                F::zero()
             } else {
-                C::Scalar::one()
+                F::one()
             },
         );
 
         Self { x, y, z }
     }
 
-    pub fn witness(cs: &mut R1cs<C>, x: C::Scalar, y: C::Scalar, is_infinity: bool) -> Self {
+    pub fn descale<C: CircuitDriver<Scalar = F>>(&self, cs: &mut R1cs<C>) -> Self {
```

Comment:
Called before using the hash function to get same values with native hash computation.
Return `(x/z, y/z, 1)` or the point if it is an identity.